### ISSUE TITLE
SSH bans despite NOFAIL prefix

### ISSUE THREAD

**Issue by microcreators** [fail2ban/fail2ban#1727]
### Environment:
- Fail2Ban version (including any possible distribution suffixes):
`Fail2Ban v0.10.0a2`
- OS, including release name/version: 
```
Linux 4.4.30-v7+ armv7l GNU/Linux
Raspbian GNU/Linux 8
```
- [ ] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:
I'm getting frequent SSH bans, due to an SSH client misbehaving and terminating connections in an abrupt way. The condition is indeed specified in the jail config, but with NOFAIL prefix.

#### Steps to reproduce
Set findtime and a number of attempts low in the jail configuration.
Connect via ssh, terminate it abruptly.

#### Expected behavior
Clients aren't banned as the F-NOFAIL is set.

#### Observed behavior
Clients are banned according to jail config.

#### Any additional information
I may be misunderstanding the concept of F-NOFAIL. Thanks for explaining.

### Configuration, dump and another helpful excerpts
sshd.conf filter line being triggered
`^<F-NOFAIL>Received disconnect</F-NOFAIL> from <HOST>: 11:`

ssh jail.local definition
```
[ssh]
enabled  = true
port     = 111
filter   = sshd
mode     = normal
logpath  = /var/log/auth.log
maxretry = 2
findtime = 5m
bantime  = 2h
```

#### Relevant lines from monitored log files in question:
Running auth.log vs the filter
```
sudo fail2ban-regex /var/log/auth.log sshd.conf

Running tests
=============

Use   failregex filter file : sshd, basedir: /etc/fail2ban
Use         maxlines : 1
Use      datepattern : Default Detectors
Use         log file : /var/log/auth.log
Use         encoding : UTF-8


Results
=======

Failregex: 25 total
|-  #) [# of hits] regular expression
|  17) [24] ^<F-NOFAIL>Received disconnect</F-NOFAIL> from <HOST>: 11:
|  18) [1] ^<F-NOFAIL>Connection closed</F-NOFAIL> by <HOST>(?: \[preauth\])?\s*$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [1050] {^LN-BEG}(?:DAY )?MON Day 24hour:Minute:Second(?:\.Microseconds)?(?: ExYear)?
`-

Lines: 1050 lines, 0 ignored, 25 matched, 1025 missed
[processed in 0.76 sec]
```

**Comment by sebres** [comment]
I don't understand what you want say us with this issue.

Indeed, both RE's ("Received disconnect" as well as "Connection closed" are specified, but they a just a helpers to get the host (IP) for the real failures (for another entries in log-file without IP address, with the same connection ID).
The fact is both above-mentioned strings may occur also by the absolutely legal ssh-connections (that make no failures at all).

In mode "normal" sshd-jail observes the authentication failures only (that you probably did not have).
Otherwise you should provide the example log-lines (or better excerpts `... | grep "sshd[connid]"` that contains such lines).

If you want use ddos, extra or aggressive modes, you should specify it e. g. as aggressive (and supply it also by `fail2ban-regex`):
```diff
- sudo fail2ban-regex /var/log/auth.log sshd.conf
+ sudo fail2ban-regex /var/log/auth.log sshd[mode=aggressive]
```
BTW I would increase `maxretry` in this case (because the value 2 by using of ddos-rules may be too small).

**Comment by microcreators** [comment]
What I was trying to say was **I'm getting banned from the server, just because of NOFAIL ssh rules. Is this the right behavior?** I don't know. There was no such concept in 0.8 :)

Examples:
Log lines triggering the ban (ssh "normal" mode only).
```
|- Matched line(s):
|  Mar 19 12:38:34 v317 sshd[9516]: Received disconnect from 94.254.131.8: 11: Bye Bye [preauth]
|  Mar 19 12:38:57 v317 sshd[9519]: Received disconnect from 94.254.131.8: 11: Bye Bye [preauth]
|  Mar 19 12:46:06 v317 sshd[9579]: Received disconnect from 95.211.230.94: 11: Bye Bye
|  Mar 19 13:37:55 v317 sshd[8636]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 19 13:49:23 v317 sshd[9255]: Received disconnect from 94.254.131.8: 11: Bye Bye [preauth]
|  Mar 19 13:49:42 v317 sshd[9264]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 19 14:15:48 v317 sshd[10477]: Received disconnect from 37.220.35.202: 11: Bye Bye
|  Mar 19 14:17:11 v317 sshd[10567]: Received disconnect from 37.220.35.202: 11: Bye Bye
|  Mar 19 15:25:36 v317 sshd[11994]: Received disconnect from 93.115.95.201: 11: Bye Bye
|  Mar 19 15:28:40 v317 sshd[12453]: Received disconnect from 93.115.95.201: 11: Bye Bye
|  Mar 19 18:00:51 v317 sshd[15936]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 19 18:17:48 v317 sshd[20774]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 19 18:52:53 v317 sshd[21694]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 19 21:06:16 v317 sshd[28262]: Received disconnect from 89.76.80.135: 11: Bye Bye
|  Mar 20 09:51:11 v317 sshd[31528]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 10:25:56 v317 sshd[2505]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 10:27:59 v317 sshd[3050]: Received disconnect from 94.254.131.8: 11:
|  Mar 20 10:33:47 v317 sshd[3139]: Connection closed by 94.254.131.8 [preauth]
|  Mar 20 12:09:45 v317 sshd[5789]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 12:30:27 v317 sshd[6304]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 15:02:51 v317 sshd[8897]: Received disconnect from 77.247.181.165: 11: Bye Bye
|  Mar 20 16:52:55 v317 sshd[11222]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 17:01:22 v317 sshd[13373]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 17:03:06 v317 sshd[14457]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 17:43:35 v317 sshd[14624]: Received disconnect from 65.19.167.130: 11: Bye Bye
|  Mar 20 18:07:02 v317 sshd[15282]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 18:15:17 v317 sshd[16576]: Received disconnect from 94.254.131.8: 11: Bye Bye
|  Mar 20 18:31:40 v317 sshd[2244]: Received disconnect from 94.254.131.8: 11: Bye Bye
```
the aggressive mode adds only one line to it:
`|  Mar 20 10:12:27 v317 sshd[639]: fatal: Read from socket failed: Connection reset by peer [preauth]`

auth log filtered using a few example connection IDs:
```
cat /var/log/auth.log | grep 9516
Mar 19 12:38:34 v317 sshd[9516]: Received disconnect from 94.254.131.8: 11: Bye Bye [preauth]
cat /var/log/auth.log | grep 9519
Mar 19 12:38:57 v317 sshd[9519]: Received disconnect from 94.254.131.8: 11: Bye Bye [preauth]
cat /var/log/auth.log | grep 9579
Mar 19 12:46:06 v317 sshd[9579]: Received disconnect from 95.211.230.94: 11: Bye Bye
```

There's nothing else suspicious that was happening. So either NO-FAIL doesn't fully work or I'm not getting the concept.

**Comment by sebres** [comment]
Hmm, strange...
Can you please provide an output of:
```
?sudo? python -c "ip='94.254.131.8'; db='/var/lib/fail2ban/fail2ban.sqlite3';  import sys, logging; logging.basicConfig(stream=sys.stdout, level=logging.ERROR); from fail2ban.server.database import Fail2BanDb; db = Fail2BanDb(db); t = db.getBansMerged(ip=ip); print(('%d attempts, matches:\n  %s' % (t.getAttempt(), '\n  '.join(t.getMatches())) ) if t else 'NOT FOUND')"
?sudo? python -c "ip='95.211.230.94'; db='/var/lib/fail2ban/fail2ban.sqlite3';  import sys, logging; logging.basicConfig(stream=sys.stdout, level=logging.ERROR); from fail2ban.server.database import Fail2BanDb; db = Fail2BanDb(db); t = db.getBansMerged(ip=ip); print(('%d attempts, matches:\n  %s' % (t.getAttempt(), '\n  '.join(t.getMatches())) ) if t else 'NOT FOUND')"
```

**Comment by sebres** [comment]
I could reproduce it, so trying to fix now...

**Comment by microcreators** [comment]
Thanks. If it helps anything now, both queries return NOT FOUND.

**Comment by sebres** [comment]
Fixed in #1728 as amend to #1719 (branch _0.10/fix-gh-1719).
I will merge it if test cases passed.

**Comment by microcreators** [comment]
Appreciate it!
